src/operation/bulk_write.rs

```diff
@@ -222,39 +233,104 @@ where
+
+            if current_size <= max_size
+                || self.encrypted && i == 0 && current_size <= first_write_max_encrypted_size
+            {
+                self.n_attempted += 1;
+                if let Some(inserted_id) = inserted_id {
+                    self.inserted_ids.insert(self.offset + i, inserted_id);
+                }
+                namespace_info.add_pending();
+                ops.push(model_document);
+            } else {
+                break;
+            }
+        }
+
+        if self.n_attempted == 0 {
+            return Err(Error::invalid_argument(format!(
+                "operation at index {} exceeds the maximum size",
+                self.offset
+            )));
+        }
+
+        let mut command = Command::new(Self::NAME, "admin", command_body);
+        namespace_info
+            .namespaces
+            .add_to_command(NS_INFO, &mut command);
+        ops.add_to_command(OPS, &mut command);
+        Ok(command)
+    }
 }
 
 /// A helper struct for tracking namespace information.
-struct NamespaceInfo<'a> {
-    namespaces: Vec<RawDocumentBuf>,
+struct NamespaceInfo<'a, T: RawDocumentCollection> {
+    namespaces: T,
+    pending_namespace: Option<RawDocumentBuf>,
     // Cache the namespaces and their indexes to avoid traversing the namespaces array each time a
     // namespace is looked up or added.
     cache: HashMap<&'a Namespace, usize>,
 }
 
-impl<'a> NamespaceInfo<'a> {
+impl<'a, T> NamespaceInfo<'a, T>
+where
+    T: RawDocumentCollection,
+{
     fn new() -> Self {
         Self {
-            namespaces: Vec::new(),
+            namespaces: Default::default(),
+            pending_namespace: None,
             cache: HashMap::new(),
         }
     }
 
-    /// Gets the index for the given namespace in the nsInfo list, adding it to the list if it is
-    /// not already present.
-    fn get_index(&mut self, namespace: &'a Namespace) -> (usize, usize) {
+    /// Gets the index for the given namespace in the nsInfo list and the number of bytes it would
+    /// add to the nsInfo list. Stores the namespace as a pending entry.
+    fn get_index_and_size(&mut self, namespace: &'a Namespace) -> Result<(usize, usize)> {
         match self.cache.get(namespace) {
-            Some(index) => (*index, 0),
+            Some(index) => Ok((*index, 0)),
             None => {
                 let namespace_doc = rawdoc! { "ns": namespace.to_string() };
-                let length_added = namespace_doc.as_bytes().len();
-                self.namespaces.push(namespace_doc);
-                let next_index = self.cache.len();
-                self.cache.insert(namespace, next_index);
-                (next_index, length_added)
+                let index = self.cache.len();
+                let bytes_added = T::bytes_added(index, &namespace_doc)?;
+                self.pending_namespace = Some(namespace_doc);
+                self.cache.insert(namespace, index);
```

Comment:
I know in current usage it doesn't matter because a pending document not being added terminates the outer loop, but it still bugs me that the namespace is added to the cache even if `add_pending` is never called.  Not sure what a good way to fix that would be, though.